packages/node/src/integrations/undici.ts

```diff
@@ -0,0 +1,231 @@
+import type { Hub, Span } from '@sentry/core';
+import { stripUrlQueryAndFragment } from '@sentry/core';
+import type { EventProcessor, Integration } from '@sentry/types';
+import { dynamicSamplingContextToSentryBaggageHeader, stringMatchesSomePattern } from '@sentry/utils';
+import type DiagnosticsChannel from 'diagnostics_channel';
+
+import type { NodeClient } from '../client';
+import { isSentryRequest } from './utils/http';
+
+enum ChannelName {
+  // https://github.com/nodejs/undici/blob/e6fc80f809d1217814c044f52ed40ef13f21e43c/docs/api/DiagnosticsChannel.md#undicirequestcreate
+  RequestCreate = 'undici:request:create',
+  RequestEnd = 'undici:request:headers',
+  RequestError = 'undici:request:error',
+}
+
+interface RequestWithSentry extends DiagnosticsChannel.Request {
+  __sentry__?: Span;
+}
+
+interface RequestCreateMessage {
+  request: RequestWithSentry;
+}
+
+interface RequestEndMessage {
+  request: RequestWithSentry;
+  response: DiagnosticsChannel.Response;
+}
+
+interface RequestErrorMessage {
+  request: RequestWithSentry;
+  error: Error;
+}
+
+interface UndiciOptions {
+  /**
+   * Whether breadcrumbs should be recorded for requests
+   * Defaults to true
+   */
+  breadcrumbs: boolean;
+}
+
+const DEFAULT_UNDICI_OPTIONS: UndiciOptions = {
+  breadcrumbs: true,
+};
+
+/** */
+export class Undici implements Integration {
+  /**
+   * @inheritDoc
+   */
+  public static id: string = 'Undici';
+
+  /**
+   * @inheritDoc
+   */
+  public name: string = Undici.id;
+
+  // Have to hold all built channels in memory otherwise they get garbage collected
+  // See: https://github.com/nodejs/node/pull/42714
+  // This has been fixed in Node 19+
+  private _channels = new Set<DiagnosticsChannel.Channel>();
+
+  private readonly _options: UndiciOptions;
+
+  public constructor(_options: UndiciOptions) {
+    this._options = {
+      ...DEFAULT_UNDICI_OPTIONS,
+      ..._options,
+    };
+  }
+
+  /**
+   * @inheritDoc
+   */
+  public setupOnce(_addGlobalEventProcessor: (callback: EventProcessor) => void, getCurrentHub: () => Hub): void {
+    let ds: typeof DiagnosticsChannel | undefined;
+    try {
+      // eslint-disable-next-line @typescript-eslint/no-var-requires
+      ds = require('diagnostics_channel') as typeof DiagnosticsChannel;
```

Comment:
Can we use `await import` here? Because of e.g. this comment:
https://github.com/getsentry/sentry-javascript/blob/develop/packages/node/src/integrations/localvariables.ts#L33

Reply:
I don't think so, because we need to support cjs.

I think that docstring is no longer an issue - but I'll let @Lms24 confirm that.

Reply:
IIRC `await import` would actually be transpiled to `Promise.resolve(() => require(..))` for our CJS build. The reason why we switched back to `require` is that only `require`'d modules  are monkey-patchable while `import`ed ones are not. So I'd say we probably can try `await import` here as we don't monkey-patch the required module. 

However, we also don't have to if it the async nature of `import` makes things harder here. For instance, I ran into timing problems when trying to convert the `LocalVariables` integration to `import` and decided to leave it as is and revisit later. 

Thanks for the comment link. I need to change the comment as I don't think it actually is true for SvelteKit anymore. I added this comment before I realized that `yarn link` caused these weird cjs/esm problems in SvelteKit.

Reply:
IMHO if we can still make it work with `import`, I'd still do it - the less `require` we have in our code base, the better!